Automattic/jetpack
# Forms: Fix fatal error due to unexpected input type

Fixes p1731428420089459-slack-C05PV073SG3

## Proposed changes:

* As far as we can see with logstash records (back to Oct 11th at initial check, but unknown how far back) there have been a few occurrences every day of a fatal error related to the `grunion_ajax_shortcode` function, on WoA sites.
* The functionality expects an array, and the proposed fix is to ensure an array is then also passed into the `usort` function.
* It isn't clear when the issue was introduced - there have been no changes in relevant files recently - however this is likely to have been an issue as long as the code was as it currently is.
* All sites of those checked did have one plugin in common - Elementor - and the issue is replicable testing with Elementor.
* The error:
```
PHP Fatal error:  Uncaught TypeError: usort(): Argument #1 ($array) must be of type array, string given in /wordpress/plugins/jetpack/14.1-a.1/jetpack_vendor/automattic/jetpack-forms/src/contact-form/class-admin.php:986
```
* Another warning was spotted while testing with Elementor related to an undefined variable, which is also fixed in this PR.

### Other information:


## Jetpack product discussion

p1731428420089459-slack-C05PV073SG3

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

**To replicate the issues:**

* Add the Elementor plugin - https://wordpress.org/plugins/elementor/ -  and activate it.
* Add add a shortcode block, then add a contact form with `[contact-form]`
* Click on the 'Edit with Elementor' button at the top of the editor.
* Click on the form in the new editor.
* In the sidebar, you should see the unrendered shortcode. Delete this. 
* Then, click on the 'Add Contact Form' button.
<img width="256" alt="Screenshot 2024-11-14 at 21 16 24" src="https://github.com/user-attachments/assets/60615505-5ce1-4711-81dc-bef994ff6acd">

* Add one or two new fields - one of them must have a field type of Email.
* Click on 'Add this form to my post' button.
<img width="363" alt="Screenshot 2024-11-14 at 19 59 53" src="https://github.com/user-attachments/assets/46ec826b-f0f4-47d2-b940-5c69587b37e2">

* Notice the fatal error in your PHP error logs, and that nothing happens with the form.

To replicate the warning:
* Clear the contact form and click 'Add Contact Form' again.
* Click 'Add this form to my post' without adding any new fields. 
* Notice the PHP warning: `Undefined variable $field_shortcodes`
* A default contact form shortcode should appear in the text editor (you may need to save fields when adding them if it's playing up).

**To test the fix:**

* Apply this PR, and follow the same steps as above. 
* You should not see a fatal error (you may see a warning regarding Array to String conversion but I believe that's related to the way we're testing, as regular forms added via the block or shortcode (also using Classic Editor) do not produce the same warning). You'll see shortcode content output on the page (again not expected in production, just how we're testing).

To test the secondary warning fix:
* Click on the 'Add this form to my post' button again but without adding any new fields. As before, a default contact form shortcode may appear in the text editor, but more importantly the warning will be gone.

**Note for WoA sites:**

* I did not see anything in the error log, but instead you should notice an error in the console instead - `Failed to load resource: the server responded with a status of 500 ()`. The fatal error is logged to Logstash though. If you use the Logstash link shared here - p1731420610837819/1731417604.113789-slack-C01U2KGS2PQ - using the current Jetpack version of 14.0-a.1 in the file filter, you should see your WoA site ID. This can be converted to the WPcom blog ID using the link further down in that Slack thread, to confirm it's your test site. 
* The functionality should work as expected though with the fix in place (and no new logs generated).
